File: Day_4/main.cpp
#include <iostream>
#include <fstream>
#include <string>

bool    contains(std::string str) //Part 1
{
    size_t pos = str.find(',');
    if (pos == std::string::npos)
    {
        std::cout << "Wrong input" << std::endl;
        return (false);
    }
    std::string s1 = str.substr(0, pos);
    std::string s2 = str.substr(pos + 1, str.length() - pos);
    
    int a1 = std::stoi(s1.substr(0, s1.find('-')));
    int a2 = std::stoi(s1.substr(s1.find('-') + 1, s1.length() - s1.find('-')));
    int b1 = std::stoi(s2.substr(0, s2.find('-')));
    int b2 = std::stoi(s2.substr(s2.find('-') + 1, s2.length() - s2.find('-')));

    if ( ( a1 <= b1 && a2 >= b2 ) || ( a1 >= b1 && a2 <= b2 ) )
        return (true);
    return (false);
}

bool overlap(std::string str) //Part 2
{
    size_t pos = str.find(',');
    if (pos == std::string::npos)
    {
        std::cout << "Wrong input" << std::endl;
        return (false);
    }
    std::string s1 = str.substr(0, pos);
    std::string s2 = str.substr(pos + 1, str.length() - pos);
    
    int a1 = std::stoi(s1.substr(0, s1.find('-')));
    int a2 = std::stoi(s1.substr(s1.find('-') + 1, s1.length() - s1.find('-')));
    int b1 = std::stoi(s2.substr(0, s2.find('-')));
    int b2 = std::stoi(s2.substr(s2.find('-') + 1, s2.length() - s2.find('-')));

    if ( ( a1 <= b1 && a2 >= b1 ) || ( a1 <= b2 && a2 >= b2 ) ||
         (b1 <= a1 && b2 >= a1) || (b1 <= a2 && b2 >= a2) )
        return (true);
    return (false);
}

int main()
{
    int res_contains = 0;
    int res_overlap = 0;
    std::string line;

    std::ifstream myfile1 ("input.txt");
    if (myfile1.is_open())
    {
        while ( getline (myfile1,line) )
        {
            if (line.length() < 7)
                continue;
            if (contains(line))
                res_contains++;
            if (overlap(line))
                res_overlap++;
        }
    }
    std::cout << "Result 1 : " << res_contains << std::endl;
    std::cout << "Result 2 : " << res_overlap << std::endl; 
    myfile1.close();

    return (0);
}

File: Day_4/main.py
def contains(str): #Part 1
    split_str = str.split(",")
    (s1, s2) = (split_str[0].split("-"))
    (s3, s4) = (split_str[1].split("-"))
    a1 = int(s1)
    a2 = int(s2)
    b1 = int(s3)
    b2 = int(s4)

    if (a1 <= b1 and a2 >= b2) or (a1 >= b1 and a2 <= b2):
        return True
    return False

def overlap(str): #Part 2
    split_str = str.split(",")
    (s1, s2) = (split_str[0].split("-"))
    (s3, s4) = (split_str[1].split("-"))
    a1 = int(s1)
    a2 = int(s2)
    b1 = int(s3)
    b2 = int(s4)

    if (a1 <= b1 and a2 >= b1) or (a1 <= b2 and a2 >= b2) or (b1 <= a1 and b2 >= a1) or (b1 <= a2 and b2 >= a2):
        return True
    return False

res_contains = int(0)
res_overlap = int(0)
with open("input.txt") as f:
    lines = f.readlines()
    for line in lines:
        if contains(line):
            res_contains += 1
        if overlap(line):
            res_overlap += 1
        
    print("Result 1 : " + str(res_contains))
    print("Result 2 : " + str(res_overlap))
